lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -95,6 +101,109 @@ def generate_replace_result_xml(result_sourcedid, score):
     return etree.tostring(xml, xml_declaration=True, encoding='UTF-8')
 
 
+def get_assignments_for_problem(problem_descriptor, user_id, course_key):
+    """
+    Trace the parent hierarchy from a given problem to find all blocks that
+    correspond to graded assignment launches for this user. A problem may
+    show up multiple times for a given user; the problem could be embedded in
+    multiple courses (or multiple times in the same course), or the block could
+    be embedded more than once at different granularities (as an individual
+    problem and as a problem in a vertical, for example).
+
+    Returns a dict where the key is a usage descriptor for a block in the
+    parent hierarchy, and the value is a list of GradedAssignment objects that
+    are associated with that descriptor for the current user.
+    """
+    assignments = defaultdict(list)
+    current_descriptor = problem_descriptor
+    while current_descriptor:
+        assignments_for_block = GradedAssignment.objects.filter(
+            user=user_id, course_key=course_key, usage_key=current_descriptor.location
+        )
+        for assignment in assignments_for_block:
+            assignments[current_descriptor].append(assignment)
+        current_descriptor = current_descriptor.get_parent()
+    return assignments
+
+
+def get_scores_for_locations(user_id, course_key):
+    """
+    Calculate the scores associated with all problems in the course. This method
+    returns a dictionary that maps a problem's usage ID to a Score object.
+    """
+    user = User.objects.get(id=user_id)
+    course = modulestore().get_course(course_key, depth=0)
+    # We make a fake request because grading code expects to be able to look at
+    # the request. We have to attach the correct user to the request before
+    # grading that student.
+    request = RequestFactory().get('/')
+    request.user = user
+    course_scores = progress_summary(user, request, course)
+    location_to_score = {}
+    for chapter in course_scores:
+        for section in chapter['sections']:
+            for score in section['scores']:
+                location_to_score[score.module_id] = score
+    return location_to_score
+
+
+def calculate_score(descriptor, location_to_score):
+    """
+    Calculate the aggregate score for a block by adding up the earned and
+    possible scores for all child blocks. Returns a tuple containing
+    (earned_score, possible_score).
+    """
+    if descriptor.location in location_to_score:
+        score = location_to_score.get(descriptor.location)
+        earned = score.earned
+        possible = score.possible
+    else:
+        earned = 0.0
+        possible = 0.0
+    for child in descriptor.get_children():
```

Comment:
This doesn't take dynamic children (e.g. A/B tests) into account. Does the information returned by `progress()` not contain this aggregated information?